edx_username_changer/signals.py

```diff
@@ -0,0 +1,45 @@
+"""
+Signals and Signal Handlers for edx-username-changer plugin
+"""
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.db.models.signals import post_save, pre_save
+from django.dispatch import receiver
+
+from common.djangoapps.util.model_utils import get_changed_fields_dict
+
+from edx_username_changer.utils import EdxUsernameChanger
+
+
+@receiver(pre_save, sender=User)
+def user_pre_save_callback(sender, **kwargs):
+    """
+    Pre-save signal handler of User model to store changed fields to be used later
+    """
+    if settings.FEATURES.get("ENABLE_EDX_USERNAME_CHANGER"):
+        user = kwargs["instance"]
+        fields_to_update = get_changed_fields_dict(user, sender)
+        if "username" in fields_to_update:
+            fields_to_update.update({"new_username": user.username})
+            user._updated_fields = fields_to_update
+
+
+@receiver(post_save, sender=User)
+def user_post_save_callback(sender, **kwargs):
+    """
+    Post-save signal handler of User model to update username throughout the application
+    """
+    if settings.FEATURES.get("ENABLE_EDX_USERNAME_CHANGER"):
+        user = kwargs["instance"]
+        if (
+            hasattr(user, "_updated_fields")
+            and user._updated_fields
+            and {"username", "new_username"}.issubset(user._updated_fields)
+        ):
+            username_changer = EdxUsernameChanger(
```

Comment:
can we encapsulate `update_user_social_auth_uid` and `update_username_in_forum` inside `EdxUsernameChanger` and pass user object and new username? if pass user object we'd not have to fetch it again inside `update_username_in_forum`.
```
username_changer = EdxUsernameChanger(user, user._updated_fields["new_username"])
username_changer.change()
```

Reply:
I have removed this extra layer of EdxUsernameChanger class, now the whole process is done through functions. Still, I couldn't send the user to the task because it was giving a `unable to serialize` error may be due to the fact that I added a temporary field.